File: examples/js/fetch-news.js
/**
 * ChainGPT AI Crypto News — Fetch & Filter Example
 *
 * Demonstrates: filtered queries, multi-category, date ranges, pagination
 * Install: npm install @chaingpt/ainews dotenv
 */
import 'dotenv/config';
import { AINews, Errors } from '@chaingpt/ainews';

const news = new AINews({ apiKey: process.env.CHAINGPT_API_KEY });

// Category IDs for reference
const CATEGORIES = {
  BLOCKCHAIN_GAMING: 2, DAO: 3, DAPPS: 4, DEFI: 5,
  LENDING: 6, METAVERSE: 7, NFT: 8, STABLECOINS: 9,
  CRYPTOCURRENCY: 64, SMART_CONTRACTS: 66, WEB3: 74, EXCHANGE: 78
};

const BLOCKCHAINS = {
  BITCOIN: 11, BNB_CHAIN: 12, ETHEREUM: 15, POLYGON: 20,
  SOLANA: 22, ARBITRUM: 28, AVALANCHE: 31, BASE: 32, OPTIMISM: 45
};

const TOKENS = {
  BTC: 79, ETH: 80, USDT: 81, BNB: 82, XRP: 83,
  SOL: 85, DOGE: 87, LINK: 96, AVAX: 92
};

// 1. Latest DeFi news
async function defiNews() {
  const res = await news.getNews({
    categoryId: [CATEGORIES.DEFI],
    limit: 5,
    sortBy: 'createdAt'
  });
  console.log('=== Latest DeFi News ===');
  res.data.forEach(article => {
    console.log(`  ${article.title}`);
    console.log(`  ${article.url}\n`);
  });
}

// 2. Ethereum-specific news about a token
async function ethTokenNews() {
  const res = await news.getNews({
    subCategoryId: [BLOCKCHAINS.ETHEREUM],
    tokenId: [TOKENS.ETH],
    limit: 10
  });
  console.log('=== ETH on Ethereum ===');
  res.data.forEach(a => console.log(`  [${a.pubDate}] ${a.title}`));
}

// 3. Search with keyword
async function searchNews(query) {
  const res = await news.getNews({
    searchQuery: query,
    limit: 10
  });
  console.log(`=== Search: "${query}" ===`);
  res.data.forEach(a => console.log(`  ${a.title} — ${a.url}`));
}

// 4. Paginated fetch (all news in last 7 days)
async function recentNewsPaginated() {
  const weekAgo = new Date(Date.now() - 7 * 24 * 60 * 60 * 1000);
  let offset = 0;
  const limit = 10;
  let total = 0;

  console.log('=== Last 7 Days (paginated) ===');
  let hasMore = true;
  while (hasMore) {
    const res = await news.getNews({
      fetchAfter: weekAgo,
      limit,
      offset
    });
    res.data.forEach(a => console.log(`  ${a.title}`));
    total += res.data.length;
    offset += limit;
    hasMore = res.data.length === limit;
  }
  console.log(`Total articles: ${total}`);
}

(async () => {
  try {
    await defiNews();
    await ethTokenNews();
    await searchNews('Bitcoin ETF');
    // await recentNewsPaginated();  // May use many credits
  } catch (error) {
    if (error instanceof Errors.AINewsError) {
      console.error('News Error:', error.message);
    } else {
      throw error;
    }
  }
})();

File: examples/python/fetch_news.py
"""
ChainGPT AI Crypto News — Fetch & Filter (Python)

Demonstrates: filtered queries, multi-category, search, pagination
Install: pip install chaingpt python-dotenv

NOTE: The Python SDK's actual class names, method signatures, and import
paths may differ from what is shown here. These examples illustrate the
*concepts* (news fetching, filtering, search) but are not guaranteed to
match the latest SDK release. Always check the official ChainGPT
documentation at https://docs.chaingpt.org for up-to-date Python SDK
usage.
"""
import asyncio
import os
from dotenv import load_dotenv
from chaingpt.client import ChainGPTClient
from chaingpt.exceptions import ChainGPTError

load_dotenv()
API_KEY = os.environ["CHAINGPT_API_KEY"]

# Reference IDs
CATEGORIES = {"DEFI": 5, "NFT": 8, "GAMING": 2, "EXCHANGE": 78, "CRYPTO": 64}
BLOCKCHAINS = {"BITCOIN": 11, "ETHEREUM": 15, "SOLANA": 22, "BNB": 12, "ARBITRUM": 28}
TOKENS = {"BTC": 79, "ETH": 80, "SOL": 85, "BNB": 82, "LINK": 96}


async def latest_defi_news():
    """Fetch latest DeFi news."""
    async with ChainGPTClient(api_key=API_KEY) as client:
        result = await client.news.get_news(
            category_id=CATEGORIES["DEFI"],
            limit=5,
            sort_by="createdAt"
        )
        print("=== DeFi News ===")
        for article in result.data:
            print(f"  {article.title}")
            print(f"  {article.url}\n")


async def search_news(query: str):
    """Search news by keyword."""
    async with ChainGPTClient(api_key=API_KEY) as client:
        result = await client.news.get_news(
            search_query=query,
            limit=10
        )
        print(f'=== Search: "{query}" ===')
        for article in result.data:
            print(f"  [{article.pubDate}] {article.title}")


async def multi_filter():
    """Ethereum DeFi news about ETH token."""
    async with ChainGPTClient(api_key=API_KEY) as client:
        result = await client.news.get_news(
            category_id=CATEGORIES["DEFI"],
            sub_category_id=[BLOCKCHAINS["ETHEREUM"]],
            token_id=TOKENS["ETH"],
            limit=10
        )
        print("=== ETH DeFi News ===")
        for a in result.data:
            print(f"  {a.title}")
        print(f"Total available: {result.total}")


async def main():
    try:
        await latest_defi_news()
        await search_news("Bitcoin ETF")
        await multi_filter()
    except ChainGPTError as e:
        print(f"Error: {e}")

if __name__ == "__main__":
    asyncio.run(main())
